mujoco_warp/_src/solver.py

```diff
@@ -101,18 +103,39 @@ def _efc_kernel(d: types.Data):
     Jaref = d.efc.Jaref[efcid]
     efc_D = d.efc.D[efcid]
 
-    # TODO(team): active and conditionally active constraints
-    active = int(Jaref < 0.0)
-    d.efc.active[efcid] = active
+    active = (Jaref < 0.0) or (efcid < d.ne[0] + d.nf[0])
+
+    if wp.static(m.dof_frictionloss_adr.size > 0 and (not DSBL_FLOSS)):
+      f = d.efc.frictionloss[efcid]
+      r = _safe_div(1.0, efc_D)
+      rf = r * f
+      f_pos = f > 0.0
+      linear_neg = (Jaref <= -rf) and f_pos
+      linear_pos = (Jaref >= rf) and f_pos
+      active = active and (not linear_neg) and (not linear_pos)
 
-    if active:
-      # efc_force = -efc_D * Jaref * active
-      d.efc.force[efcid] = -1.0 * efc_D * Jaref
+      linear_negf = float(linear_neg) * f
+      linear_posf = float(linear_pos) * f
+      floss_force = linear_negf - linear_posf
 
-      # cost = 0.5 * sum(efc_D * Jaref * Jaref * active))
-      wp.atomic_add(d.efc.cost, worldid, 0.5 * efc_D * Jaref * Jaref)
+      floss_cost = linear_negf * (-0.5 * rf - Jaref)
+      floss_cost += linear_posf * (-0.5 * rf + Jaref)
+
+      cost = floss_cost
+      efc_force = floss_force
     else:
-      d.efc.force[efcid] = 0.0
+      cost = 0.0
+      efc_force = 0.0
+
+    # TODO(team): benchmark options for active
```

Comment:
what does this TODO mean?

Reply:
this todo is referring to benchmarking the performance of multiplying by `active` versus `active` in an if statement.

the code is updated to if statements and this todo has been removed.